client/services/LocalStorageService.js

```diff
@@ -0,0 +1,214 @@
+       * @param {string} questionId
+       * @param {string} language
+       * @returns {string|null}
+       */
+      loadStoredCode: function(questionId, language) {
+        if (!localStorageIsAvailable) {
+          return null;
+        }
+
+        var localStorageKey = getLocalStorageKeyForCode(
+          questionId, language);
+        var storedCodeWithVersion = localStorage.getItem(localStorageKey);
+
+        if (storedCodeWithVersion === null) {
+          return null;
+        }
+        // TODO(talee): Check versioning when we start having different versions
+        var storedCode = storedCodeWithVersion.substring(
+          storedCodeWithVersion.indexOf(':') + 1);
+
+        return storedCode;
+      },
+
+      /**
+       * Clears the local storage such that there is no longer any code
+       * stored there.
+       *
+       * @param {string} questionId
+       * @param {string} language
+       */
+      clearLocalStorageCode: function(questionId, language) {
+        if (!localStorageIsAvailable) {
+          return;
+        }
+
+        var localStorageKey = getLocalStorageKeyForCode(
+          questionId, language);
+        localStorage.removeItem(localStorageKey);
+      },
+
+      /**
+       * Takes the paragraphs and reinforcement bullets and creates an object
+       * that is then converted to json and stored.
+
+       * @param {string} questionId
+       * @param {Array} feedback
+       * @param {Array} reinforcement
+       * @param {string} language
+       */
+      storeFeedback: function(questionId, feedback, reinforcement, language) {
+        if (!localStorageIsAvailable) {
+          return;
+        }
+
+        var localStorageKey = getLocalStorageKeyForFeedback(
+          questionId, language);
+        var feedbackWithReinforcement = {};
+        feedbackWithReinforcement.feedbackParagraphs = feedback;
+        feedbackWithReinforcement.reinforcementBullets = reinforcement;
+
+        localStorage.setItem(localStorageKey,
+          angular.toJson(feedbackWithReinforcement));
+      },
+
+      /**
+       * Loads the feedback and parses it into JSON, and then reconstructs
+       * the feedback paragraphs & reinforcement bullets fron the JSON object
+       *
+       * @param {string} questionId
+       * @param {string} language
+       * @returns {Object}
+       */
+      loadStoredFeedback: function(questionId, language) {
+        if (!localStorageIsAvailable) {
+          return null;
+        }
+
+        var localStorageKey = getLocalStorageKeyForFeedback(
+          questionId, language);
+        var storedFeedback = localStorage.getItem(localStorageKey);
+        if (storedFeedback === null) {
+          return null;
+        }
+        var rawFeedback = angular.fromJson(storedFeedback);
+
+        var reconstructedFeedback = [];
+        for (var i = 0; i < rawFeedback.feedbackParagraphs.length; i++) {
+          reconstructedFeedback = rawFeedback.feedbackParagraphs.map(
```

Comment:
Why do you still need the 'j' for loop? You seem to be constructing the variable `reconstructedFeedback` multiple times.